src/Grpc.AspNetCore.Web/Internal/GrpcWebMiddleware.cs

```diff
@@ -0,0 +1,157 @@
+﻿#region Copyright notice and license
+
+// Copyright 2019 The gRPC Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#endregion
+
+using System;
+using System.Threading.Tasks;
+using Grpc.Shared;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Features;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Grpc.AspNetCore.Web.Internal
+{
+    internal sealed class GrpcWebMiddleware
+    {
+        private readonly GrpcWebOptions _options;
+        private readonly ILogger<GrpcWebMiddleware> _logger;
+        private readonly RequestDelegate _next;
+
+        public GrpcWebMiddleware(IOptions<GrpcWebOptions> options, ILogger<GrpcWebMiddleware> logger, RequestDelegate next)
+        {
+            _options = options.Value;
+            _logger = logger;
+            _next = next;
+        }
+
+        public Task Invoke(HttpContext httpContext)
+        {
+            var mode = GetGrpcWebMode(httpContext);
+            if (mode != GrpcWebMode.None)
+            {
+                Log.DetectedGrpcWebRequest(_logger, httpContext.Request.ContentType);
+
+                var metadata = httpContext.GetEndpoint()?.Metadata.GetMetadata<IGrpcWebEnabledMetadata>();
+                if (metadata?.GrpcWebEnabled ?? _options.GrpcWebEnabled)
+                {
+                    return HandleGrpcWebRequest(httpContext, mode);
+                }
+
+                Log.GrpcWebRequestNotProcessed(_logger);
+            }
+
+            return _next(httpContext);
+        }
+
+        private async Task HandleGrpcWebRequest(HttpContext httpContext, GrpcWebMode mode)
+        {
+            var feature = SetGrpcWebFeature(mode, httpContext);
```

Comment:
I had considered that. Because the gRPC-Web middleware changes the content type, if the gRPC-Web middleware runs again then it will no-op.

Are there other reasons to undo setting the features? Or is it considered something you should do to be a good middleware citizen?

Reply:
> Are there other reasons to undo setting the features? Or is it considered something you should do to be a good middleware citizen?

Both. In the error handling scenario the route will change so you'll end up at a different endpoint. If the base64 PipeWriter is still in place that could get ugly. Reverting also avoids side-effects for middleware earlier in the pipeline.

Reply:
Done